include/javabind/enum.hpp

```diff
@@ -0,0 +1,97 @@
+/**
+ * javabind: effective C++ and Java interoperability
+ * @see https://github.com/hunyadi/javabind
+ *
+ * Copyright (c) 2024 Levente Hunyadi
+ *
+ * This work is licensed under the terms of the MIT license.
+ * For a copy, see <https://opensource.org/licenses/MIT>.
+ */
+
+#pragma once
+#include "exception.hpp"
+#include "message.hpp"
+#include "object.hpp"
+#include "signature.hpp"
+#include <algorithm>
+#include <optional>
+#include <type_traits>
+
+namespace javabind
+{
+    template <typename native_type>
+    struct EnumValues
+    {
+        inline static std::map<native_type, std::string_view> values_to_names;
+        inline static std::map<std::string_view, native_type> names_to_values;
+
+        static native_type native_value(std::string_view java_name)
+        {
+            return names_to_values.at(java_name);
+        }
+
+        static std::string_view java_name(native_type native_value)
+        {
+            return values_to_names.at(native_value);
+        }
+
+        static bool contains(native_type native_value)
+        {
+            return values_to_names.find(native_value) != values_to_names.end();
+        }
+
+        static bool contains(std::string_view java_name)
+        {
+            return names_to_values.find(java_name) != names_to_values.end();
+        }
+
+        static void bind(native_type native_value, std::string_view java_name)
+        {
+            values_to_names[native_value] = java_name;
+            names_to_values[java_name] = native_value;
+        }
+    };
+
+    template <typename T>
+    struct EnumClassJavaType : AssignableJavaType<T>
+    {
+        using native_type = T;
+        using java_type = jobject;
+
+        using AssignableJavaType<T>::class_path;
+        using AssignableJavaType<T>::class_name;
+        using AssignableJavaType<T>::sig;
+
+        static native_type native_value(JNIEnv* env, java_type javaEnumValue)
+        {
+            LocalClassRef enumClass(env, javaEnumValue);
+            Method nameFunc = enumClass.getMethod("name", FunctionTraits<std::string()>::sig);
+            jobject nameObject = env->CallObjectMethod(javaEnumValue, nameFunc.ref());
+            std::string nameString = arg_type_t<std::string>::native_value(env, static_cast<jstring>(nameObject));
```

Comment:
We could consider calling `ordinal()` instead of `name()`, and map numeric indices in Java to enumeration values in C++. This might be more efficient if many Java/C++ enumeration value conversions take place. Java itself seems to make use of `ordinal()` in `EnumMap` and `EnumSet`.

Reply:
Changed the implementation to store an `std::unordered_map` of ordinals to native types to resolve a java enum object to the native type